### PR TITLE
Add iOS build support to the CMake build

### PR DESCRIPTION
**Pull request by wilhuff**
Individual commits more-or-less say what they're doing.

At a high level, this change:

  * Allows a user to disable building tests and benchmarks
  * Fixes iOS-specific problems that arise from cross-compilation difficulty with CMake
  * Undoes the special case in #5042, adding a general `DISABLE_STRICT_WARNINGS` flag and removing the `-Wno-foo` blocks that were subtracting from the full warnings battery. This is ultimately a better choice for effectively third-party code.

Note:
  * this is diffbased on #5051;
  * I'll add a separate PR to add CI for this

### CODE REVIEW

**Comment by wilhuff** [comment]
Diffbase removed.

File: CMakeLists.txt [hunk 1]
```diff
@@ -98,19 +104,32 @@ download_external_sources()
...
-set(BENCHMARK_ENABLE_INSTALL OFF CACHE BOOL "Firestore disabled")
-set(BENCHMARK_ENABLE_GTEST_TESTS OFF CACHE BOOL "Firestore disabled")
-add_external_subdirectory(benchmark)
+if(FIREBASE_IOS_BUILD_BENCHMARKS)
+  set(BENCHMARK_ENABLE_TESTING OFF CACHE BOOL "Firestore disabled")
+  set(BENCHMARK_ENABLE_EXCEPTIONS OFF CACHE BOOL "Firestore disabled")
+  set(BENCHMARK_ENABLE_INSTALL OFF CACHE BOOL "Firestore disabled")
+  set(BENCHMARK_ENABLE_GTEST_TESTS OFF CACHE BOOL "Firestore disabled")
+
+  if(IOS)
```
Comment (a-maurice) [inline CMakeLists.txt]: FYI, IOS is only a standard CMake variable as of version 3.14.  For the C++ repo, we had to explicitly define IOS via the toolchain file to get around this (https://github.com/firebase/firebase-cpp-sdk/blob/master/cmake/toolchains/ios.cmake#L44), though will probably work on changing that to use a more explicit TARGET_PLATFORM_IOS, or something similar.  You might want to do something similar as well, to support older cmakes.

File: cmake/cc_rules.cmake [hunk 2]
```diff
@@ -225,33 +233,59 @@ function(firebase_ios_maybe_remove_objc_sources output_var)
...
-# Adds FIREBASE_IOS_OBJC_FLAGS to the compile options of the given target if
-# any of the sources have filenames that indicate they are Objective-C.
-function(firebase_ios_add_objc_flags target)
-  set(_has_objc OFF)
+# If any of the sources have filenames that indicate they are Objective-C adds
+# Either FIREBASE_IOS_OBJC_FLAGS or FIREBASE_IOS_OBJC_FLAGS_STRICT depending on
+# whether or not DISABLE_STRICT_WARNINGS was passed.
+function(firebase_ios_add_compile_options target)
+  set(flag DISABLE_STRICT_WARNINGS)
+  cmake_parse_arguments(aco "${flag}" "" "${}" ${ARGN})
```
Comment (a-maurice) [inline cmake/cc_rules.cmake]: Is there a reason for the empty brackets?

File: CMakeLists.txt (hunk 1 above)
Comment (wilhuff) [reply CMakeLists.txt]: The "standard" ios toolchain file also sets `IOS`: https://github.com/leetal/ios-cmake/blob/master/ios.toolchain.cmake#L466

Coding to modern CMake and adding fallback support for older CMake versions seems like a reasonable path forward. If/when `firebase-cpp-sdk` chooses to implement some other convention, we can revisit here too.

File: cmake/cc_rules.cmake (hunk 2 above)
Comment (wilhuff) [reply cmake/cc_rules.cmake]: No. This is a mistake (though I'm surprised it works). Fixed.

**Comment by wilhuff** [comment]
Travis failures are unrelated.

**Pull request merged**